python/cuda_cccl/cuda/compute/struct.py

```diff
@@ -183,154 +386,127 @@ def tuple_to_struct_cast(context, builder, fromty, toty, val):
+    @lower_cast(types.Tuple, StructType)
+    @lower_cast(types.UniTuple, StructType)
+    def cast_tuple_to_struct(context, builder, fromty, toty, val):
+        """Lower implicit cast from tuple to struct."""
+        # Check tuple size matches struct fields
+        if isinstance(fromty, types.UniTuple):
+            if fromty.count != len(field_spec):
+                return None
+            tuple_types = [fromty.dtype] * fromty.count
         else:
-            getitem_lines.append(f"    elif idx == {i}:")
-        getitem_lines.append(f"        return self.{field_name}")
-
-    getitem_lines.append("    else:")
-    getitem_lines.append(
-        f"        raise IndexError(f'Index {{idx}} out of range for struct with {len(field_names)} fields')"
-    )
+            if len(fromty.types) != len(field_spec):
+                return None
+            tuple_types = list(fromty.types)
 
-    getitem_code = f"""
-def __getitem__(self, idx):
-    '''Get field by index.'''
-{chr(10).join(getitem_lines)}
-"""
-
-    # Generate __len__ method
-    len_code = f"""
-def __len__(self):
-    '''Return the number of fields.'''
-    return {len(field_names)}
-"""
+        # Build struct from tuple elements, casting each to the correct field type
+        struct_val = cgutils.create_struct_proxy(toty)(context, builder)
+        for i, (field_name, field_type) in enumerate(field_spec.items()):
+            elem = builder.extract_value(val, i)
+            # Cast tuple element to field type
+            elem = context.cast(builder, elem, tuple_types[i], field_type)
+            setattr(struct_val, field_name, elem)
 
-    # Execute the generated code and add methods to the class
-    namespace: Dict[str, Any] = {}
-    exec(getitem_code, namespace)
-    exec(len_code, namespace)
+        return struct_val._getvalue()
 
-    StructClass.__getitem__ = namespace["__getitem__"]  # type: ignore
-    StructClass.__len__ = namespace["__len__"]  # type: ignore
+    return struct_class
 
-    # Set up the class for use with numba
-    _setup_numba_struct(StructClass, field_dict)
 
-    return StructClass
+def _patch_struct_class(struct_class):
+    """Add the required Python-side attributes to the struct class after creation."""
 
+    def __init__(self, *args, **kwargs):
+        """Supporting construction from positional, keyword, and dict arguments."""
+        field_spec = self.__class__._field_spec
 
-def gpu_struct(this: type) -> Type[GpuStruct]:
-    """
-    Decorate a class as a GPU struct.
+        # Normalize to dict
+        if args and isinstance(args[0], dict):
+            values = args[0]
+        elif args:
+            assert len(args) == len(field_spec), (
+                f"Expected {len(field_spec)} arguments, got {len(args)}"
+            )
+            values = dict(zip(field_spec.keys(), args))
+        else:
+            values = kwargs
+
+        assert values.keys() == field_spec.keys()
```

Comment:
Suggestion: the branching logic here is hard to follow. I'm not sure if it can be improved, but it would definitely help to move this to a new method, maybe `get_fields_as_dict` or something more appropriate.